lib/usagereporter/daemon/usagereporter.go

```diff
@@ -132,7 +131,8 @@ func GetAnonymizedPrehogEvent(req *teletermv1.ReportUsageEventRequest) (*prehogv
 	}
 
 	// anonymized
-	anonymizer, err := newClusterAnonymizer(req.GetAuthClusterId()) // authClusterId is sent with each event that needs to be anonymized
+	// TODO: update connect code to set this correctly (cluster ID normally, anonymization key when it's set in the license)
+	anonymizer, err := utils.NewHMACAnonymizer(req.GetAnonymizationKey()) // anonymizationKey is sent with each event that needs to be anonymized
```

Comment:
Question: what is this connect code mentioned here?

I can see that this request object is formed client-side here: https://github.com/gravitational/teleport/blob/f17ace0041fa35eefa544e037e321c86de4d7d8d/web/packages/teleterm/src/services/tshd/mapUsageEvent.ts#L25-L32

However, not sure how make anonymization key to become available here.

Reply:
@ravicious or @gzdunek can probably help here

Reply:
I'm about to log off so I didn't have time to familiarize myself with how the anonymization key is obtained from the license. But I suspect it follows the same rule as `authClusterId`, we have to ask the cluster for it. I hope that explaining how we get `authClusterId` will help with what you need and I'll make sure to answer your questions when I start work tomorrow.

---

A _very_ brief intro to the Electron app architecture is here:
* https://github.com/gravitational/teleport/blob/9c4fa206c9868c61052d6a13c50a64c63cdea602/lib/teleterm/README.md
* https://github.com/gravitational/teleport/blob/9c4fa206c9868c61052d6a13c50a64c63cdea602/web/packages/teleterm/README.md#diagram

When the Electron app starts, it asks the tsh daemon for the list of all root clusters. This list is returned based on the state on disk, think `tsh status`. Once it gets the list, it issues a request to each cluster for full cluster details. This is [`Handler.GetCluster`](https://github.com/gravitational/teleport/blob/9c4fa206c9868c61052d6a13c50a64c63cdea602/lib/teleterm/apiserver/handler/handler_clusters.go#L82) in tshd. The response is stored in [`ClustersService`](https://github.com/gravitational/teleport/blob/9c4fa206c9868c61052d6a13c50a64c63cdea602/web/packages/teleterm/src/ui/services/clusters/clustersService.ts#L508-L509) in the frontend app of the Electron app. The response includes `authClusterId`, that's how we're able to [include it with every event](https://github.com/gravitational/teleport/blob/9c4fa206c9868c61052d6a13c50a64c63cdea602/web/packages/teleterm/src/ui/services/usage/usageService.ts#L248-L251) sent from the Electron app to tshd.

Assuming that the anonymization key can be obtained from the auth server, you should be able to return it the same way we return auth cluster ID. You need to update [the `Cluster` message in protos](https://github.com/gravitational/teleport/blob/9c4fa206c9868c61052d6a13c50a64c63cdea602/proto/teleport/lib/teleterm/v1/cluster.proto#L26-L50), run `make grpc`, then follow the trail in `Handler.GetCluster` to return the anonymization key from there somehow. Once that's done, what's left is updating `getClusterProperties` in `usageService.ts` so that it returns that anonymization key as well.

We don't use `authClusterId` for anything else, so perhaps instead of returning both `authClusterId` and `anonymizationKey` in the `Cluster` message, it'd make sense to return just `anonymizationKey` that sometimes happens to be an auth cluster ID. That assumes that both things are scoped per root cluster.

Reply:
I looked at gravitational/teleport.e#2872 and I see that the anonymization key at the moment is available only in teleport.e. Connect uses tsh underneath and there's no enterprise version of tsh, so tsh will need the ability to somehow get the anonymization key from the cluster.

@gzdunek and @zmb3 should be able to provide more guidance here as they were recently discussing adding support for another enterprise feature which supporting code was available only in teleport.e but now needs to be accessible by the tsh daemon as well.

Reply:
I agree with @ravicious, we should replace `authClusterId` with `anonymizationKey` in the `Cluster` message.

As for getting the key from the license, I think we can retrieve it from the cluster's `authClient`:
```go
pem, err := authClient.GetLicense(ctx)
if err != nil {
	return nil, trace.Wrap(err)
}
licenseFile, err := licensefile.FromPEM([]byte(pem))
if err != nil {
	return nil, trace.Wrap(err)
}

licenseFile.License.GetAnonymizationKey()
```

Reply:
Thanks a lot for clarification.

I _renamed_ `auth_cluster_id` to `anonymization_key` in proto files. As it is only used to be sent back as is, this change should be completely transparent for Teleport Connect.

Reply:
Thanks for the explanation. Changed to reserve + new field.

Reply:
@ravicious if the teleterm protos are 100% guaranteed to only be used between matching versions of teleport connect and tsh daemon we can remove them from the backwards compatibility checker, FYI

(if they're not then we can't remove `auth_cluster_id` from one major version to the next, that'd have to wait until Teleport 16 lol)

Reply:
They are, but I remember Zac mentioning that it's better to just enforce consistent rules across all protos in the project rather than having an exception where we can break the rules. People would have remember about the exception or be reminded of it etc. So generally keeping the lint on for lib/teleterm seems like a sensible approach to me.